### PR TITLE
#2378 - Overapproximation of a paving

### LINKED ISSUE #2378: Overapproximation of implicit set using ICP + polyhedral outer approximation
The idea is transform an (implicitly defined) nonlinear constraint and compute a union of polyhedra where the constraint is guaranteed to be verified (and it is complete).

Note: there is an additional complexity if the set has disconnected components. In a first stage we can assume that there is only 1 connected component.

### PR DESCRIPTION
**Pull request by mforets**
Closes #2378.

Reference notebook:  https://nbviewer.jupyter.org/github/mforets/escritoire/blob/master/2020/Week41/ICP_OA.ipynb

- [x] add docstring

- [x] add test

### CODE REVIEW

**Review by mforets — COMMENTED** [review]
i can't approve my own PR, but lgtm

File: src/Approximations/overapproximate.jl [hunk 1]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+
+
+"""
+    overapproximate(p::Paving{L, N}, dirs::AbstractDirections{N, VN}) where {L, N, VN}
+
+Overapproximation of a Paving-type set representation using an HPolytope.
+
+### Input
+
+- `p`    -- type of pave representation
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1986]:
```suggestion
- `p`    -- paving

File: src/Approximations/overapproximate.jl [hunk 2]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+- `dirs` -- directions of the HPolytope
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1987]:
```suggestion
- `dirs` -- template directions
```

File: src/Approximations/overapproximate.jl [hunk 3]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+
+### Output
+
+An overapproximation of a Paving-type set using an HPolytope with constraints in direction dirs.
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1991]:
```suggestion
An overapproximation of a paving using a polyhedron in constraint representation (`HPolyhedron`) with constraints in direction `dirs`.
```

File: src/Approximations/overapproximate.jl [hunk 4]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+
+### Algorithm
+
+This function takes the union of the elements in the boundary of p, first
+converted into hyperrectangles, and then calculates the support function of the
+set along each  direction in dirs, to compute the HPolytope constraints.
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1997]:
```suggestion
set along each  direction in dirs, to compute the `HPolyhedron` constraints.
```

File: test/unit_overapproximate.jl [hunk 5]
```diff
@@ -421,5 +421,16 @@ for N in [Float64]
         L = Line2D(N[-1, -1], N[1, 1/2])
         cap = overapproximate(Z ∩ L, OctDirections)
         @test (L ∩ Z) ⊆ cap
+
+        # overapproximate of a pave with an HPolytope
```
Comment (mforets) [inline test/unit_overapproximate.jl:425]:
```suggestion
        # overapproximate a nonlinear constraint with an HPolyhedron
```

File: src/Approximations/overapproximate.jl [hunk 6]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+    Uouter = UnionSetArray(convert.(Hyperrectangle, p.boundary))
+    constraints = [HalfSpace(d, ρ(d, Uouter)) for d in dirs]
+    return HPolyhedron(constraints)
+end
+
+# alias with HPolyhedron type as second argument
+"""
+    overapproximate(p::Paving{L, N}, ::Type{<:HPolyhedron}, dirs::AbstractDirections{N, VN}) where {L, N, VN}
+
+Dispath in
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:2012]:
```suggestion
Dispatch to
```

File: src/Approximations/overapproximate.jl [hunk 7]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+
+### Algorithm
+
+This function takes the union of the elements in the boundary of p, first
+converted into hyperrectangles, and then calculates the support function of the
+set along each  direction in dirs, to compute the HPolytope constraints.
+
+### Requires IntervalConstraintProgramming
+"""
+function overapproximate(p::Paving{L, N}, dirs::AbstractDirections{N, VN}) where {L, N, VN}
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:2001]: this function needs to be added to the docs..?

File: test/unit_overapproximate.jl [hunk 8]
```diff
@@ -421,5 +421,16 @@ for N in [Float64]
...
+        if N == Float64
+            dom = IntervalBox(IntervalArithmetic.Interval(-2, 2), IntervalArithmetic.Interval(-2, 2))
+            C = IntervalConstraintProgramming.@constraint x^2 + y^2 <= 1
+            p = IntervalConstraintProgramming.pave(C, dom, 0.01)
+            dirs = OctDirections(2)
+            H = overapproximate(p, dirs);B2 = Ball2(N[0., 0.], N(1.))
```
Comment (schillic) [inline test/unit_overapproximate.jl:431]:
```suggestion
            H = overapproximate(p, dirs)
            B2 = Ball2(N[0, 0], N(1))
```

File: src/Approximations/overapproximate.jl [hunk 9]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+    # enclose outer approximation
+    Uouter = UnionSetArray(convert.(Hyperrectangle, p.boundary))
+    constraints = [HalfSpace(d, ρ(d, Uouter)) for d in dirs]
+    return HPolyhedron(constraints)
+end
+
+# alias with HPolyhedron type as second argument
+"""
+    overapproximate(p::Paving{L, N}, ::Type{<:HPolyhedron}, dirs::AbstractDirections{N, VN}) where {L, N, VN}
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:2010]: same here, though i think it's fine to remove this docstring altogether

File: src/Approximations/overapproximate.jl (hunk 7 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl:2001]: Not in this case. We just let `Documenter` build [all docs](https://github.com/JuliaReach/LazySets.jl/blob/master/docs/src/lib/approximations.md#overapproximations) for the function `overapproximate`.

File: src/Approximations/overapproximate.jl [hunk 10]
```diff
@@ -1963,3 +1963,59 @@ function overapproximate(X::Intersection{N, <:Hyperplane, <:AbstractZonotope},
...
+
+return quote
+
+using .IntervalConstraintProgramming: Paving
+
+
+"""
+    overapproximate(p::Paving{L, N}, dirs::AbstractDirections{N, VN}) where {L, N, VN}
+
+Overapproximation of a Paving-type set representation using an HPolytope.
```
Comment (schillic) [inline src/Approximations/overapproximate.jl:1982]:
```suggestion
Overapproximation of a Paving-type set representation using a polyhedron in constraint representation.
```

**Review by schillic — APPROVED** [review]

**Review by schillic — APPROVED** [review]

**Pull request merged**